Limelight Vision

JacobAaCook opened the pull request.

Commit:
Limelight Vision start

Commit:
Worked on Drive

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -0,0 +1,58 @@
+    posY = posData[1];
+    dist = posData[2];
+
+    double turnRequest = posX * TURN_K;
```
You'll likely want to use MoPID for these since you'll want to specify targets (e.g. 0 for X and some positive value for dist).  You'll also likely want the PID control bundled up instead of just a naked P.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,63 @@
+   *
+   */
+
+  public double validTarget = NetworkTableInstance.getDefault().getTable("limelight").getEntry("tv").getDouble(0);
```
This statement is executed once at class construction time.  Thus, the tv entry will only be read once and validTarget will be whatever it is when the robot boots.
Same goes for xOffset and yOffset.

It's ok to create NetworkTableEntry members at class construction, but you will need to getDouble periodically, or you will need to attach listeners to the Entries.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,63 @@
+  public double xOffset = NetworkTableInstance.getDefault().getTable("limelight").getEntry("tx").getDouble(0);
+  public double yOffset = NetworkTableInstance.getDefault().getTable("limelight").getEntry("ty").getDouble(0);
+
+  public static final double RANGE_X = 29.8;
```
I'm assuming you pulled these values from the Limelight example.  You should document what they mean and why they were chosen.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,63 @@
+  private static final double TARGET_ANGLE = 0; // a2
+  private double height = TARGET_HEIGHT - CAMERA_HEIGHT;
+  private double angle = Math.tan(CAMERA_ANGLE + TARGET_ANGLE);
+  public double targetDistance = (height) / (angle);
```
As above, these values are only being calculated once, ever.  If that's the intent, then fine.  But otherwise this math needs to be in a method that gets called periodically.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  protected void execute() {
+    posData = limelight.targetPos();
```
limelight.targetPos can return null.  You must validate that posData is non-null before using it below.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,63 @@
+   * detects whether or not the Limelight has detected a valid target and returns
+   * the position of the target in an array of [x, y, z]
+   */
+  public double[] targetPos() {
```
returning bundled values as an array is a really bad code smell.  There is no implicit checking for the size of the array.  You are restricted to only values of a single type.  The caller has to "just know" what each of the slots in the array represents.  There are hideous long-term maintenance issues.

Just don't do this.

It's far better to create an immutable class that holds all the values with correct names and types.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/Limelight.java:
```diff
@@ -0,0 +1,63 @@
+    }
+  }
+
+  public boolean atTarget() {
```
Having targetIsValid, targetPos, and atTarget broken out as separate methods opens you up to multithreaded race conditions.  You have no guarantees that the internal values won't change between calls so you can't know if the call to targetIsValid actually relates to the data returned from targetPos.

The way this should be handled is that the Limelight subsystem has one synchronized call to get a snapshot of the current state as an immutable object.  That object then has methods to determine if it's valid and if so, what position the target is at.

Commit:
Restructuring 

I moved the Data read from the limelight into its own class and moved both the Limelight and LimelightData classes into their own sub-folder

JacobAaCook committed:
Made More Changes from PR

I think I got rid of the race error possibilities caused by breaking up the validTarget, TargetPos and atTarget methods.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -10,38 +10,31 @@
+      double x = Math.abs(data.xCoord());
+      double y = Math.abs(data.yCoord());
+      double d = data.dist();
+      if (x <= X_ERR && y <= Y_ERR && d + DIST_ERR <= TARGET_DIST || d - DIST_ERR >= TARGET_DIST) {
```
The DIST_ERR calculation should be done as  abs(difference) <= constant, just like x and y

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/vision/LimelightData.java:
```diff
@@ -0,0 +1,56 @@
+
+import edu.wpi.first.networktables.*;
+
+public class LimelightData {
```
I don't think you quite understood what I meant by an "immutable class".  We should discuss tonight about the flow of data, the lifetime of the different class objects and why they should be split that way.

terbos90803 replied:
This class was folded into Limelight and this file should be deleted.

JacobAaCook committed:
More work on the PR Changes

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -5,13 +5,17 @@
   private LimelightData data = Robot.limelightData;
 
-  public double posX;
+  public static final double RANGE_X = 29.8; // The range of values given by "tx" is from -29.8 to 29.8 degrees
```
These values are specific to the Limelight and thus should be located in the Limelight class.

Always think about who "owns" a piece of information.  All ownership should be clear and localized.  It's ok to ask another class for a piece of information, like a constant, as long as the "truth" is owned by the correct class.

This is called the "Information Expert" principle.  A specific class should always be the owner of all of its related data.  Any other class that needs that data needs to ask the expert instead of having its own copy.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -20,21 +23,85 @@
+      return validTarget;
+    }
+
+    public boolean targetMet(){
```
This doesn't appear to be complete.  The implementation of targetMet needs to move from the out class into here.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/subsystems/vision/Limelight.java:
```diff
@@ -4,14 +4,17 @@
+  private final NetworkTableEntry validEntry = table.getEntry("tv");
 
-  public double validTarget = NetworkTableInstance.getDefault().getTable("limelight").getEntry("tv").getDouble(0);
+  public LimelightData data;
```
There is no reason for Limelight to ever hold onto a LimelightData.  Remove this member.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -32,11 +36,11 @@ protected void initialize() {
 
   protected void execute() {
-    posX = data.xCoord();
+    turnErr = data.xCoord();
```
The first thing that needs to happen in execute() is to get a fresh LimelightData.  Make that call and hold the data as a local variable in this method.

JacobAaCook committed:
Condensed and Finished up some methods

Fixed/Finished the targetMet() method. 
Deleted the LimelightData file, since it has become a nested class. 
Added the part where the VisionDrive actually grabs new data

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -21,7 +21,7 @@
 import frc.robot.subsystems.HatchPassive;
 import frc.robot.subsystems.vision.Limelight;
-import frc.robot.subsystems.vision.LimelightData;;
+import frc.robot.subsystems.vision.Limelight.LimelightData;
```
Extra import?  Remove if not needed.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -36,6 +36,7 @@ protected void initialize() {
   }
 
   protected void execute() {
+    limelight.getData();
```
LimelightData data = limelight.getData();

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -4,13 +4,13 @@
   private DriveSubsystem drive = Robot.driveSystem;
   private Limelight limelight = Robot.limelight;
-  private LimelightData data = Robot.limelightData;
+  private LimelightData data;
```
LimelightData should not be a member of VisionDrive.  Remove.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -36,6 +36,7 @@ protected void initialize() {
 
   protected void execute() {
+    limelight.getData();
     turnErr = data.xCoord();
```
All of these variables should be local to this method.  There is no reason for most of these to be class members.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -47,7 +48,7 @@ protected void execute() {
   @Override
   protected boolean isFinished() {
-    return limelight.targetMet();
+    return data.targetMet();
```
There does need to be some way to connect the data retrieved in execute to the answer to the isFinished question.

While it's possible to hold data as a class member, I would recommend instead calling targetMet in execute and holding the result as a boolean class member.  Then just return that boolean here.

JacobAaCook replied:
so have a separate boolean in VisionDrive? Is there any reason why to, the way I see it is, targetMet() will return True or False, if we have a separate boolean that will just echo whatever targetMet() is saying. So it seems like more steps than should be needed. Unless doing it this way is some cause for a race condition?

terbos90803 replied:
The structure of a Command (VisionDrive extends Command) is that is has an execute() method and an isFinished() method.  These two methods are called separately, but need to be consistent in their actions/responses.  To provide this consistency, there needs to be some connection between the two.  This is best handled by using private members of the class, which are shared between the two methods.

So you have some choices to make about the form of those shared members.  For this situation, you have two basic options:

1. get a fresh copy of LimelightData on every call to execute and do *some* of the processing on that data there (i.e. drive commands).  Hold onto that LimelightData as a class member and do additional processing on it during the call to isFinished (i.e. targetMet).
1. get a fresh copy of LimelightData on every call to execute and do *all* processing on that data there, discarding the LimelightData when execute exits.  Save the results of targetMet as a boolean class member to be used during the anticipated future call to isFinished.

Either one of these options will achieve the desired result, but the preference is to choose option 2 for these reasons:

- In general, as much processing as possible should be handled in the execute class.  The designers of the Command pattern expected this call to be more expensive than any other since it should be doing all the work.
- isFinished is effectively an accessor.  It should not be doing any calculations that should have been done during execute.  All it needs to do is reference the results of those calculations and return that pre-computed boolean.

While it's unlikely in this specific situation, you should always expect accessors to be called multiple times per execute cycle with the expectation that they return the same result as quickly as possible.  So as a good practice, the accessor should do as little processing as possible, which most frequently is just returning a stored value.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -0,0 +1,62 @@
+    turnErr = data.xCoord();
+    dist = data.dist();
+
+    double turnRequest = Utils.map(turnErr, -RANGE_X, RANGE_X, -1.0, 1.0);
```
You haven't checked data.isValid to see if the coordinates are valid.

You need to make sure that the data is valid and useful before you act on it.  If you don't have valid data, the best choice is to force all movement requests to 0 for safety.

JacobAaCook committed:
Checked for validity of positions and made targetMet() a class member

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -10,16 +10,15 @@
 
-  public static final double RANGE_X = 29.8; // The range of values given by "tx" is from -29.8 to 29.8 degrees
-  public static final double RANGE_Y = 24.85; // The range of values given by "ty" is from -24.85 to 24.85 degrees
+  public static final double RANGE_X = Limelight.RANGE_X; // The range of values given by "tx" is from -29.8 to 29.8
```
While it may seem convenient to have these local copies of the Limelight values, there is no penalty in simply using the Limelight class values directly when needed.  Doing so reduces the amount of redundant clutter here.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -10,16 +10,15 @@
-  public double turnErr;
-  public double posY;
-  public double dist;
   public double limelightTurn;
```
Are these unused?  If so, remove.

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -36,19 +35,27 @@ protected void initialize() {
-    double turnRequest = Utils.map(turnErr, -RANGE_X, RANGE_X, -1.0, 1.0);
-    double moveRequest = Utils.map(dist, -RANGE_Y, RANGE_Y, -1.0, 1.0);
+    LimelightData data = limelight.getData();
+    boolean valid = data.valid();
```
Why do you make local copies of all of these values?  There is little to no penalty in calling accessors, so it's more concise to just call them inline where needed, especially if they're only used once.
The biggest justifications for making a local copy is if you are trying to address:

- performance across multiple uses
- readability in the case of very long calling strings, particularly when using multiple times
- readability if the name of the accessor is quite different from the intended local usage

In this case, valid and valid are a match and only used once, so should just be inline.
turnErr and xCoord are sufficiently different that you could argue in favor of making the local to enhance readability.

terbos90803 reviewed (changes requested):
Getting really close.  Down to the last few nits.

JacobAaCook committed:
Removed local instances of variables called only once

terbos90803 commented on 2019RobotCodeDeepSpace/src/main/java/frc/robot/commands/VisionDrive.java:
```diff
@@ -36,16 +31,13 @@ protected void initialize() {
-    double dist = data.dist();
     double turnRequest;
     double moveRequest;
     met = data.targetMet();
```
met needs to be qualified by valid as well.  Can't be at the target if you can't see it.

terbos90803 reviewed (changes requested):
Pretty sure this is the last one.

JacobAaCook committed:
forgot to check validity before saying we are at the target
